core/lib/zksync_core/src/tee_verifier_input_producer/mod.rs

```diff
@@ -0,0 +1,262 @@
+use zksync_object_store::{ObjectStore, ObjectStoreFactory};
+use zksync_prover_interface::inputs::PrepareBasicCircuitsJob;
+use zksync_queued_job_processor::JobProcessor;
+use zksync_state::{PostgresStorage, ReadStorage, StorageView};
+use zksync_tee_verifier::TeeVerifierInput;
+use zksync_types::{L1BatchNumber, L2ChainId};
+use zksync_utils::u256_to_h256;
+
+use self::metrics::METRICS;
+
+mod metrics;
+
+/// Component that extracts all data (from DB) necessary to run a TEE Verifier.
+#[derive(Debug)]
+pub struct TeeVerifierInputProducer {
+    connection_pool: ConnectionPool<Core>,
+    l2_chain_id: L2ChainId,
+    object_store: Arc<dyn ObjectStore>,
+}
+
+impl TeeVerifierInputProducer {
+    pub async fn new(
+        connection_pool: ConnectionPool<Core>,
+        store_factory: &ObjectStoreFactory,
+        l2_chain_id: L2ChainId,
+    ) -> anyhow::Result<Self> {
+        Ok(TeeVerifierInputProducer {
+            connection_pool,
+            object_store: store_factory.create_store().await,
+            l2_chain_id,
+        })
+    }
+
+    fn process_job_impl(
+        rt_handle: Handle,
+        l1_batch_number: L1BatchNumber,
+        started_at: Instant,
+        connection_pool: ConnectionPool<Core>,
+        object_store: Arc<dyn ObjectStore>,
+        l2_chain_id: L2ChainId,
+    ) -> anyhow::Result<TeeVerifierInput> {
+        let prepare_basic_circuits_job: PrepareBasicCircuitsJob = rt_handle
+            .block_on(object_store.get(l1_batch_number))
+            .context("failed to get PrepareBasicCircuitsJob from object store")?;
+
+        let mut connection = rt_handle
+            .block_on(connection_pool.connection())
+            .context("failed to get connection for TeeVerifierInputProducer")?;
+
+        let miniblocks_execution_data = rt_handle.block_on(
+            connection
+                .transactions_dal()
+                .get_l2_blocks_to_execute_for_l1_batch(l1_batch_number),
+        )?;
+
+        let last_batch_miniblock_number = miniblocks_execution_data.first().unwrap().number - 1;
+
+        let l1_batch_header = rt_handle
+            .block_on(connection.blocks_dal().get_l1_batch_header(l1_batch_number))
+            .with_context(|| format!("header is missing for L1 batch #{l1_batch_number}"))?
+            .unwrap();
+
+        let l1_batch_params_provider = rt_handle
+            .block_on(L1BatchParamsProvider::new(&mut connection))
+            .context("failed initializing L1 batch params provider")?;
+
+        let first_miniblock_in_batch = rt_handle
+            .block_on(
+                l1_batch_params_provider
+                    .load_first_l2_block_in_batch(&mut connection, l1_batch_number),
+            )
+            .with_context(|| {
+                format!("failed loading first miniblock in L1 batch #{l1_batch_number}")
+            })?
+            .with_context(|| format!("no miniblocks persisted for L1 batch #{l1_batch_number}"))?;
```

Comment:
Nit/bikeshedding: this impl has a lot of `block_on` which looks a bit odd. Maybe it would be prettier to fetch all the data in `Self::process_job` (which is async) and pass here via structure? I think it may make the code here less verbose.

Reply:
I tried the simple approach with using `tokio::task::spawn(async move { ... })` in `process_job()`, but that clashes with `PostgresStorage::load_factory_dep()` doing a `tokio::runtime::Handle::block_on()`, which will result in a tokio panic on runtime.

Reply:
found a workaround